### DISCUSSION TITLE
Why not protobufs?

### DISCUSSION THREAD

**Discussion by maxzaver** [near/borsh#110] [Q&A]
Pasting an answer from Slack here:

Protobufs optimize for interoperability at expense of having the following issues:
* They impose extra representation language. If you have a type A in Rust/C++/Python that uses types that do not map perfectly to Protobuf types then we need to implement an extra conversion step, meaning we would have to to from binary blob to protobuf object to object of type A.
* The same protobuf object might have several valid representations as binary blob which is a source for bugs;
* Optimizing for interop creates complexity which makes it slow: https://github.com/erickt/rust-serialization-benchmarks (deserialization is 2x slower than bincode. And bincode deserialization is about 6x slower than borsh for NEAR blocks);
* Optimizing for interop makes it code complex. Rust protobuf implementation is 45k LOC with tests, while borsh is 4.5k LOC.

We can fix issues with protobufs being non-deterministic by adding various checks and converters, e.g. see https://github.com/tendermint/go-amino/ but this will make it even more complex and slow. And serialization speed and code size are extremely important for out networking and Wasm execution, it is one of the major bottlenecks. There is good reason why Mozilla is using https://github.com/servo/bincode instead of protobufs -- it is much faster and simpler. Borsh is basically https://github.com/servo/bincode with less code (it is not using serde), fixed determinism, and faster. bincode and borsh are 99% compatible.

**Comment by artob** [comment]
@MaksymZavershynskyi For posterity, is the origin story of Borsh documented or recounted anywhere? As in, who came up with it originally (which might conceivably be different from who pushed the first commit)?
